autoroutes.pyx

```diff
@@ -47,16 +65,51 @@ cdef class Edge:
             str rest = self.pattern[prefix_len:]
         new_child.connect(self.child, rest)
         self.child = new_child
+        new_child.compile()
         self.pattern = self.pattern[:prefix_len]
 
+    cdef Node join(self, str path):
+        cdef:
+            unsigned int local_index, candidate_index
+            Edge candidate = Edge(path, None)
+        candidate.compile()
+        local_index, candidate_index = self.compare(candidate)
+        del candidate
+        if not local_index:
+            return None
+        if local_index < self.pattern_len:
+            self.branch_at(local_index)
+        if candidate_index < len(path):
+            return self.child.insert(path[candidate_index:])
+        return self.child
+
+    cpdef tuple compare(self, Edge other):
```

Comment:
Is the `cpdef` required, I don't think we use it in Python?

Reply:
`other` looks weird, maybe keep `candidate` or `challenger` or something like that

Reply:
humm, now that we have split from the `join` part, I feel like the other is not candidate to anything in the `compare` method, it's just an other edge we want to compare to, and feel this is the [python](https://github.com/python/cpython/blob/8b57d7363916869357848e666d03fa7614c47897/Lib/collections/__init__.py#L283) way of calling.

Reply:
Which makes me think we can use special comparison methods from Python but that might be for another iteration ;)